syntax = "proto3";

package distributed_system;

service NodeService {
  rpc SendRequest(RequestMessage) returns (ReplyMessage);
  
  rpc SendAccept(AcceptMessage) returns (AcceptedMessage);
  rpc SendCommit(CommitMessage) returns (Empty);
  
  rpc SendPrepare(PrepareMessage) returns (PreparedMessage);
  rpc SendCommit2PC(Commit2PCMessage) returns (Empty);
  rpc SendAbort2PC(Abort2PCMessage) returns (Empty);
  rpc SendAcknowledgment(AcknowledgmentMessage) returns (Empty);
  
  rpc SendFail(FailMessage) returns (Empty);
  rpc SendRecover(RecoverMessage) returns (Empty);
  rpc SendLeaderFailed(FailMessage) returns (Empty);
  
  rpc SendNewView(NewViewMessage) returns (Empty);
  rpc SendPrepareElection(PrepareElectionMessage) returns (PromiseMessage);
  
  rpc GetBalance(BalanceRequest) returns (BalanceResponse);
  rpc GetAllBalances(Empty) returns (AllBalancesResponse);
  rpc Flush(Empty) returns (Empty);
  rpc SendReshard(ReshardMessage) returns (ReshardResponse);
}


message Empty {}


message Transaction {
  int32 sender = 1;
  int32 receiver = 2;
  int32 amount = 3;
  int32 item_id = 4;
}


message RequestMessage {
  int32 sender = 1;
  string transaction_id = 2;
  Transaction transaction = 3;
}

message ReplyMessage {
  int32 sender = 1;
  string transaction_id = 2;
  string status = 3;
  int32 balance = 4;
}


message AcceptMessage {
  int32 sender = 1;
  int32 ballot = 2;
  int32 seq = 3;
  PaxosValue value = 4;
  string phase = 5;
}

message AcceptedMessage {
  int32 sender = 1;
  int32 ballot = 2;
  int32 seq = 3;
}

message CommitMessage {
  int32 sender = 1;
  int32 seq = 2;
  PaxosValue value = 3;
}

message PaxosValue {
  string transaction_id = 1;
  Transaction transaction = 2;
  string type = 3;
  string phase = 4; 
  int32 coordinator = 5; 
}


message PrepareMessage {
  int32 sender = 1;
  string transaction_id = 2;
  Transaction transaction = 3;
}

message PreparedMessage {
  int32 sender = 1;
  string transaction_id = 2;
  string status = 3;
}

message Commit2PCMessage {
  int32 sender = 1;
  string transaction_id = 2;
}

message Abort2PCMessage {
  int32 sender = 1;
  string transaction_id = 2;
}

message AcknowledgmentMessage {
  int32 sender = 1;
  string transaction_id = 2;
}


message FailMessage {
  int32 sender = 1;
  int32 node_id = 2;
}

message RecoverMessage {
  int32 sender = 1;
  int32 node_id = 2;
}

message NewViewMessage {
  int32 sender = 1;
  int32 view = 2;
  int32 cluster = 3;
}

message PrepareElectionMessage {
  int32 sender = 1;
  int32 ballot = 2;
  int32 cluster = 3;
}

message PromiseMessage {
  int32 sender = 1;
  int32 ballot = 2;
  bool promised = 3;
  int32 last_accepted_ballot = 4;
  repeated UncommittedValue uncommitted_values = 5;
}

message UncommittedValue {
  int32 seq = 1;
  int32 ballot = 2;
  PaxosValue value = 3;
}


message BalanceRequest {
  int32 item_id = 1;
}

message BalanceResponse {
  int32 balance = 1;
}

message BalanceEntry {
  int32 item_id = 1;
  int32 balance = 2;
}

message AllBalancesResponse {
  repeated BalanceEntry balances = 1;
}

message ReshardMessage {
  int32 sender = 1;
}

message ReshardRecord {
  int32 item_id = 1;
  int32 source_cluster = 2;
  int32 destination_cluster = 3;
}

message ReshardResponse {
  repeated ReshardRecord moved_records = 1;
}
